plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -227,29 +222,122 @@ func (g *genericScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister,
 	if err != nil {
 		return nil, nil, nil, err
 	}
-	for len(nodeToVictims) > 0 {
-		node := pickOneNodeForPreemption(nodeToVictims)
-		if node == nil {
+
+	// scheduler extenders that do not support preemption
+	unPreemptableExtenders := []algorithm.SchedulerExtender{}
+
+	if len(nodeToVictims) > 0 {
+		for _, extender := range g.extenders {
+			if extender.SupportsPreemption() {
+				// Pass nodeToVictims to extender to do extra preemption logics. We can expect it returns a nodeToVictimsAfterExtender struct with:
+				//   1. Subset of the given nodes after Filter.
+				//   2. Additional victims on the nodes after preemption.
+				if nodeToVictimsAfterExtender, err := extender.ProcessPreemption(pod, nodeToVictims, g.cachedNodeInfoMap, pdbs); err != nil {
+					return nil, nil, nil, err
+				} else {
+					// Write back result after extender to nodeToVictims.
+					for nodeName := range nodeToVictims {
+						// If node exists in nodeToVictims but does not exist in nodeToVictimsAfterExtender, delete it.
+						if victimAfterExtender, ok := nodeToVictimsAfterExtender[nodeName]; !ok {
+							delete(nodeToVictims, nodeName)
+						} else {
+							nodeToVictims[nodeName] = victimAfterExtender
+						}
+					}
+				}
+			} else {
+				unPreemptableExtenders = append(unPreemptableExtenders, extender)
+			}
+		}
+
+		nodeName := pickOneNodeForPreemption(nodeToVictims)
+		if len(nodeName) == 0 {
 			return nil, nil, nil, err
 		}
-		passes, pErr := nodePassesExtendersForPreemption(pod, node.Name, nodeToVictims[node].pods, g.cachedNodeInfoMap, g.extenders)
-		if passes && pErr == nil {
-			// Lower priority pods nominated to run on this node, may no longer fit on
-			// this node. So, we should remove their nomination. Removing their
-			// nomination updates these pods and moves them to the active queue. It
-			// lets scheduler find another place for them.
-			nominatedPods := g.getLowerPriorityNominatedPods(pod, node.Name)
-			return node, nodeToVictims[node].pods, nominatedPods, err
+
+		// Lower priority pods nominated to run on this node, may no longer fit on
+		// this node. So, we should remove their nomination. Removing their
+		// nomination updates these pods and moves them to the active queue. It
+		// lets scheduler find another place for them.
+		nominatedPods := g.getLowerPriorityNominatedPods(pod, nodeName)
```

Comment:
This should be executed when we have found the node for preemption. So, please delete this block from here and also from the block below. We can save the final node in a variable and run this for the node after we are done with the extenders.

Reply:
Good catch, just fixed the PR and also extract common logic to functions.